src/core/lib/security/credentials/xds/xds_credentials.cc

```diff
@@ -31,11 +34,101 @@ constexpr const char XdsCredentials::kCredentialsTypeXds[];
+bool VerifySingleSubjectAlterativeName(
+    const std::string& subject_alternative_name, const std::string& matcher) {
+  if (subject_alternative_name.empty() ||
+      absl::StartsWith(subject_alternative_name, ".")) {
+    // Illegal pattern/domain name
+    return false;
+  }
+  if (matcher.empty() || absl::StartsWith(matcher, ".")) {
+    // Illegal domain name
+    return false;
+  }
+  // Normalize sanToVerify and pattern by turning them into absolute domain
+  // names if they are not yet absolute. This is needed because server
+  // certificates do not normally contain absolute names or patterns, but they
+  // should be treated as absolute. At the same time, any sanToVerify presented
+  // to this method should also be treated as absolute for the purposes of
+  // matching to the server certificate.
+  std::string normalized_san =
+      absl::EndsWith(subject_alternative_name, ".")
+          ? subject_alternative_name
+          : absl::StrCat(subject_alternative_name, ".");
+  std::string normalized_matcher =
+      absl::EndsWith(matcher, ".") ? matcher : absl::StrCat(matcher, ".");
+  absl::AsciiStrToLower(&normalized_san);
+  absl::AsciiStrToLower(&normalized_matcher);
+  if (!absl::StrContains(normalized_san, "*")) {
+    return normalized_san == normalized_matcher;
+  }
+  // WILDCARD PATTERN RULES:
+  // 1. Asterisk (*) is only permitted in the left-most domain name label and
+  // must be the
+  //    only character in that label (i.e., must match the whole left-most
+  //    label). For example, *.example.com is permitted, while *a.example.com,
+  //    a*.example.com, a*b.example.com, a.*.example.com are not permitted.
+  // 2. Asterisk (*) cannot match across domain name labels.
+  //    For example, *.example.com matches test.example.com but does not match
+  //    sub.test.example.com.
+  // 3. Wildcard patterns for single-label domain names are not permitted.
+  if (!absl::StartsWith(normalized_san, "*.")) {
+    // Asterisk (*) is only permitted in the left-most domain name label and
+    // must be the only character in that label
+    return false;
+  }
+  if (normalized_san == "*.") {
+    // Wildcard pattern for single-label domain name -- not permitted.
+    return false;
+  }
+  std::string suffix = normalized_san.substr(1);
+  if (absl::StrContains(suffix, "*")) {
+    // Asterisk (*) is not permitted in the suffix
+    return false;
+  }
+  if (!absl::EndsWith(normalized_matcher, suffix)) return false;
+  int suffix_start_index = normalized_matcher.length() - suffix.length();
+  // Asterisk matching across domain labels is not permitted.
+  return suffix_start_index <= 0 /* should not happen */ ||
+         normalized_matcher.find_last_of('.', suffix_start_index - 1) ==
+             std::string::npos;
+}
+
+bool VerifySubjectAlternativeNames(const char* const* subject_alternative_names,
```

Comment:
As discussed offline, it would be great if we can move this function to a common-shared file(like `tls_utils.h`), so that other classes can reuse it in the future.

Reply:
Thanks for putting the verification function to a common place!

Reply:
looks like reviewable did not post the response here